x-pack/filebeat/input/httpjson/request.go

```diff
@@ -34,6 +35,41 @@ type httpClient struct {
 	limiter *rateLimiter
 }
 
+type idsStruct struct {
+	ids []string
+	err error
+}
+
+type processResponse struct {
+	httpResponse *http.Response
+	idsReady     *sync.Mutex
+}
+
+func block(resp *http.Response) processResponse {
+	var mu sync.Mutex
+	mu.Lock()
+	return processResponse{resp, &mu}
+}
+
+func (p *processResponse) block() processResponse {
+	var mu sync.Mutex
+	mu.Lock()
+	p.idsReady = &mu
+	return *p
+}
+
+func (p processResponse) done() {
+	if p.idsReady != nil {
+		p.idsReady.Unlock()
+	}
+}
+
+func (p processResponse) wait() {
+	if p.idsReady != nil {
+		p.idsReady.Lock()
+	}
+}
```

Comment:
```suggestion
type processResponse struct {
	httpResponse *http.Response
	
	// idsReady ensures that a response will not be processed
	// until double chain calls responses have been finalized.
	// While idsReady is a sync.Mutex it is being used here as
	// a waitgroup with a single member. This means that the
	// wait method may only be called once, otherwise a deadlock
	// will result. If multiple wait calls are required in the future,
	// the type should be replaced with a sync.WaitGroup.
	idsReady *sync.Mutex
}

// block returns a processResponse that will block until response completion.
func block(resp *http.Response) processResponse {
	var mu sync.Mutex
	mu.Lock()
	return processResponse{resp, &mu}
}

// block upgrades the receiver to a blocking state. It may only be called once
// in the lifetime of the processResponse. The upgraded value is returned.
func (p *processResponse) block() processResponse {
	var mu sync.Mutex
	mu.Lock()
	p.idsReady = &mu
	return *p
}

// done releases the receiver block. It may only be called once in the lifetime
// of the processResponse.
func (p *processResponse) done() {
	if p.idsReady != nil {
		p.idsReady.Unlock()
	}
}

// wait will block until the receiver's block has been released. In the current
// implementation it may only be called once in the lifetime of the processResponse.
func (p *processResponse) wait() {
	if p.idsReady != nil {
		p.idsReady.Lock()
	}
}
```